### PR TITLE
Ingest: Add script processor

### PR DESCRIPTION
**Pull request by spinscale**
This adds another processor, that supports scripting for fields.
Right now it works by using the return value of a script to update
a specific field, as this blends better into the ingest processor infrastructure
than updating a full document.

Another possibility would be to have the same as the Update API, where you
can set arbitrary fields in the map and then store this map as the new source.

Current example call:

```
curl -X PUT localhost:9200/_ingest/pipeline/script-pipeline -d '
{
  "description" : "script pipeline",
  "processors" : [
    {
    "script" : {
      "field" : "bytes_total",
      "lang" : "painless",
      "inline" : "return input.ctx._source.doc.bytes_in + input.ctx._source.doc.bytes_out"
    }
  }
 ]
}
'

curl -X POST localhost:9200/_ingest/pipeline/script-pipeline/_simulate -d '
{
  "docs" : [
    { "_source" : { "bytes_in" : 1024, "bytes_out" : 2048 } }
  ]
}
'
```

TO DISCUSS: Alternatively behave like the Update API and allow to configure anything. However this is much
harder to implement, as the IngestDocument only allows for single field updates and a complete replacement
of the ingest document would be more tricky and resource intensive.

```
curl -X PUT localhost:9200/_ingest/pipeline/script-pipeline -d '
{
  "description" : "script pipeline",
  "processors" : [
    {
      "script" : {
        "lang" : "painless",
        "inline" : "input.ctx._source.bytes_total = input.ctx._source.doc.bytes_in + input.ctx._source.doc.bytes_out"
      }
    }
  ]
}
'
```

### CODE REVIEW

File: docs/plugins/ingest.asciidoc [hunk 1]
```diff
@@ -21,6 +21,13 @@ This processor adds this information by default under the `geoip` field.
 The ingest-geoip plugin ships by default with the GeoLite2 City and GeoLite2 Country geoip2 databases from Maxmind made available
 under the CCA-ShareAlike 3.0 license. For more details see, http://dev.maxmind.com/geoip/geoip2/geolite2/. 
 
+<<ingest-script>>::
+
+The script processor allows you to have scripted fields when indexing documents. So you could easily add two fields or do complex calculations
+beofre a document is indexed, instead of doing this on query time.
```
Comment (nik9000) [inline docs/plugins/ingest.asciidoc]: s/beofre/before/

File: plugins/ingest-script/src/main/java/org/elasticsearch/ingest/script/IngestScriptPlugin.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    }
+
+    @Override
+    public String description() {
+        return "Ingest processor that supports scripting";
+    }
+
+    public void onModule(NodeModule nodeModule) throws IOException {
+        // cant get the scriptservice/clusterservice here directly, still being initiated
+        // must be done deferredly in the factory
```
Comment (nik9000) [inline plugins/ingest-script/src/main/java/org/elasticsearch/ingest/script/IngestScriptPlugin.java]: s/be done deferredly/deferred until construction time/ ?

File: plugins/ingest-script/src/main/java/org/elasticsearch/ingest/script/ScriptProcessor.java [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+        private Node node;
+
+        public Factory(Node node) {
+            this.node = node;
+        }
+
+        @Override
+        protected ScriptProcessor doCreate(String processorTag, Map<String, Object> config) throws Exception {
+            String field = readStringProperty(TYPE, processorTag, config, "field");
```
Comment (nik9000) [inline plugins/ingest-script/src/main/java/org/elasticsearch/ingest/script/ScriptProcessor.java]: We already have script parsing code and this seems to duplicate some of it?

File: plugins/ingest-script/src/main/java/org/elasticsearch/ingest/script/ScriptProcessor.java [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+                script = new Script(file, FILE, lang, emptyMap());
+            } else if (Strings.hasLength(inline)) {
+                script = new Script(inline, INLINE, lang, emptyMap());
+            } else if (Strings.hasLength(id)) {
+                script = new Script(id, STORED, lang, emptyMap());
+            } else {
+                throw new ElasticsearchException("Could not initialize script");
+            }
+
+            ScriptService scriptService = node.injector().getInstance(ScriptService.class);
```
Comment (nik9000) [inline plugins/ingest-script/src/main/java/org/elasticsearch/ingest/script/ScriptProcessor.java]: Yikes! I'd really prefer not to do this. This kind of thing is hacky when you like guice and it is super bad when you are trying to leave guice. I'm not really sure the right thing here though.

File: plugins/ingest-script/src/main/java/org/elasticsearch/ingest/script/ScriptProcessor.java [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+        this.clusterService = clusterService;
+    }
+
+    @Override
+    public void execute(IngestDocument ingestDocument) throws Exception {
+        Map<String, Object> data = new HashMap<>();
+        Map<String, Object> context = new HashMap<>();
+        data.put("ctx", context);
+        context.put("_source", ingestDocument.getSourceAndMetadata());
+        CompiledScript compiledScript = scriptService.compile(script, ScriptContext.Standard.INGEST, emptyMap(), clusterService.state());
```
Comment (nik9000) [inline plugins/ingest-script/src/main/java/org/elasticsearch/ingest/script/ScriptProcessor.java]: If you wanted to so something like _update you could implement a Map out of _source and IngestDocument I think. We did/do those kind of things with fields.

File: plugins/ingest-script/src/main/java/org/elasticsearch/ingest/script/ScriptProcessor.java (hunk 3 above)
Comment (talevy) [inline plugins/ingest-script/src/main/java/org/elasticsearch/ingest/script/ScriptProcessor.java]: this `readStringProperty`? it is not parsing script code, it is an ingest-node config reader

File: plugins/ingest-script/src/main/java/org/elasticsearch/ingest/script/ScriptProcessor.java (hunk 4 above)
Comment (talevy) [inline plugins/ingest-script/src/main/java/org/elasticsearch/ingest/script/ScriptProcessor.java]: I agree. This dependency should be passed into the `ScriptProcessor.Factory`

I don't think this is better: https://github.com/elastic/elasticsearch/pull/18193/files#diff-509dbced5cd63de8c1bc2c13a829af0dR83

but I just piggyback off of the TemplateService.

So, we still need to figure out the best way to do this

File: plugins/ingest-script/src/main/java/org/elasticsearch/ingest/script/ScriptProcessor.java (hunk 5 above)
Comment (talevy) [inline plugins/ingest-script/src/main/java/org/elasticsearch/ingest/script/ScriptProcessor.java]: maybe I am missing something, but `.getSourceAndMetadata()` returns a mutable Map?

here is an example:
https://github.com/elastic/elasticsearch/pull/18193/files#diff-4e27382bea1f95bce321ce30c5315e98R42

**Comment by talevy** [comment]
We need to combine our efforts here: https://github.com/elastic/elasticsearch/pull/18193 :)

The first question I have is whether this should be a plugin, module, or in core?

**Comment by rmuir** [comment]
btw if you merge up, the painless syntax has changed in master. you don't need `input.ctx` just `ctx` (and if you need to access a parameter, its eg. `params.param1`). `ctx` is a reserved word for the executable scripts (e.g. updates), IMO the right way to go here. adding something named `doc` ala #18193 is not going to work, because its reserved for the search api.

**Comment by spinscale** [comment]
Closing this one in favor of #18193 after syncing with @martijnvg

**Pull request closed without merging**